cpu/nrf52/periph/spi.c

```diff
@@ -0,0 +1,240 @@
+}
+
+static void _clear_workaround(spi_t bus)
+{
+    NRF_PPI->CHENCLR = 1U << spi_config[bus].ppi;
+}
+#endif
+
+void spi_init(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+
+    /* initialize mutex */
+    mutex_init(&locks[bus]);
+    mutex_init(&busy[bus]);
+    mutex_lock(&busy[bus]);
+    /* initialize pins */
+    spi_init_pins(bus);
+}
+
+void spi_init_pins(spi_t bus)
+{
+    gpio_init(spi_config[bus].sclk, GPIO_OUT);
+    gpio_init(spi_config[bus].mosi, GPIO_OUT);
+    gpio_init(spi_config[bus].miso, GPIO_IN);
+    /* select pins for the SPI device */
+    SPI_SCKSEL  = spi_config[bus].sclk;
+    SPI_MOSISEL = spi_config[bus].mosi;
+    SPI_MISOSEL = spi_config[bus].miso;
+#ifdef CPU_MODEL_NRF52832XXAA
+    _setup_workaround_for_ftpan_58(bus);
+#endif
+    spi_twi_irq_register_spi(dev(bus), spi_isr_handler, (void*)bus);
+}
+
+int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
+{
+    (void) cs;
+
+    mutex_lock(&locks[bus]);
+    /* configure bus */
+    dev(bus)->CONFIG = mode;
+    dev(bus)->FREQUENCY = clk;
+    /* enable the bus */
+    dev(bus)->ENABLE = SPIM_ENABLE_ENABLE_Enabled;
+
+    return SPI_OK;
+}
+
+void spi_release(spi_t bus)
+{
+    /* power off everything */
+    dev(bus)->ENABLE = 0;
+    mutex_unlock(&locks[bus]);
+}
+
+static void _transfer(spi_t bus, const uint8_t *out_buf, uint8_t *in_buf, uint8_t transfer_len)
+{
+    uint8_t out_len = (out_buf) ? transfer_len : 0;
+    uint8_t in_len = (in_buf) ? transfer_len : 0;
+    const uint8_t *out_mbuf = out_buf;
+    /**
+     * Copy the out buffer in case it resides in flash, EasyDMA only works from
+     * RAM
+     */
+    if ((out_buf < (uint8_t*)SPI_CPU_FLASH_END) && (out_len)) {
+        memcpy(_mbuf[bus], out_buf, out_len);
+        out_mbuf = _mbuf[bus];
+    }
+    dev(bus)->TXD.PTR = (uint32_t)out_mbuf;
+    dev(bus)->RXD.PTR = (uint32_t)in_buf;
+
+    dev(bus)->TXD.MAXCNT = out_len;
+    dev(bus)->RXD.MAXCNT = in_len;
+
+    /* clear any spurious END events */
+    dev(bus)->EVENTS_END = 0;
+    dev(bus)->TASKS_START = 1;
+}
+
+void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
+                        const void *out, void *in, size_t len)
+{
+    const uint8_t *out_buf = out;
+    uint8_t *in_buf = in;
+
+    assert(out_buf || in_buf);
+
+    if (cs != SPI_CS_UNDEF) {
+        gpio_clear((gpio_t)cs);
+    }
+
+    /* Enable the workaround when the length is only 1 byte */
+#ifdef CPU_MODEL_NRF52832XXAA
+    size_t _len = len;
+    if (_len == 1) {
+        _enable_workaround(bus);
+    }
+#endif
+
+    /* Enable IRQ */
+    dev(bus)->INTENSET = SPIM_INTENSET_END_Msk;
+
+    do {
+        size_t transfer_len = len > UINT8_MAX ? UINT8_MAX : len;
+        _transfer(bus, out_buf, in_buf, transfer_len);
+        /* Block until the irq releases the mutex, then lock it again for the
+         * next transfer */
+        mutex_lock(&busy[bus]);
+        out_buf += out_buf ? transfer_len : 0;
+        in_buf += in_buf? transfer_len : 0;
```

Comment:
```suggestion
        in_buf += in_buf ? transfer_len : 0;
```

Reply:
I should probably give the whole file a round of uncrustify

Reply:
And done